### PR TITLE
Add No Escape Final Traitor Objective (aka Singularity Shuttle Event)

### PR DESCRIPTION
**Pull request by timothymtorres**
## About The Pull Request
This is a remake of:
 
- #77188 
- #86655

Both were DNM'd due to a lack of difficulty requirements for spawning a singularity as a shuttle event.

---

**No Escape - Final Traitor Objective:**

- Spawns a special singularity beacon, syndicate inducer, and wrench.
- The beacon must be powered with an inducer and planted on the shuttle to work.
- The beacon slowly increases the chance of a massive STAGE SIX (11x11) singularity to appear (1% every  8 seconds)
- The beacon can be turned on at any time, but will only increase the chance while on the shuttle and if it is in transit
- After 5 seconds the crew gets an announcement that a singularity is approaching and has an extra minute of transit time due to time dilation
- If the beacon is turned off or destroyed it decreases the probability by the same rate. (-1% every 8 seconds)
- If the beacon is spaced while active it decreases the probability by x2 rate. (-2% every 8 seconds)
- If the singularity is spawned while the beacon is disabled or spaced, there is a chance for it to not directly hit the shuttle (but since it's so big it will likely brush against the side)

To prevent the singularity from instantly appearing and to give the crew a chance to react, it starts with a negative probability that takes 15 seconds to reach 0%. Deactivating, destroying, or spacing the beacon will slowly reverse the chance but it's not an instant guarantee. So the longer you wait to act, the worse your chances are!

I cleaned up quite a bit of the singularity code while I was working on this.
CC @!Time-Green  @!MrMelbert 

## Why It's Good For The Game
There have been several attempts to add a singularity shuttle event that could be triggered but it was deemed too chaotic or the requirements too easy so they were restricted to admin-only events. Making it a final traitor objective, sets a high requirement that must be achieved before activating it as a doomsday event. It also gives the crew a chance to intervene and stop the event before disaster strikes. 

It's similar to a syndicate bomb ticking down while on the shuttle that serves to be climatic.

## Changelog
:cl:
add: Add no escape final traitor objective that spawns a stage six (11x11) singularity shuttle event.
/:cl:

### CODE REVIEW

**Comment by Cheshify** [comment]
More traitor final objectives that just encourage sec to kill every antag they can before they turn into a nigh unstoppable force. I dislike this PR because it further contributes towards that.

**Comment by zoomachina** [comment]
only 1% every 8 seconds??

**Comment by SmArtKar** [comment]
> More traitor final objectives that just encourage sec to kill every antag they can before they turn into a nigh unstoppable force. I dislike this PR because it further contributes towards that.

That's kind of what already happens though. I doubt that this will change much, or anything, in terms of sec behavior but it does introduce a bit of variety and can spice up evacuations in ways other than bombing or flooding the shuttle

**Comment by EnterTheJake** [comment]
I don't really see the point in adding a singularity final objectie when we already have the dark matter singularity in the game.

**Comment by DaydreamIQ** [comment]
This...really doesn't seem like something that fun to have as a final objective. Usually the final objectives are methods that CAUSE a shuttle call, but having an entire mechanic dedicated to essentially just EORG is just boring. At that point just plant a syndicate bomb and blow everyone else up, also 1% for every 8 seconds? So if you have phenomenally bad rng or someone else hijacks you get screwed out of the one thing you spent all round building up to

**Comment by timothymtorres** [comment]
Keep in mind the probability is cumulative. To make it simple if your rate of change was 1% then every second your chance of the event being triggered would be:
```
1 seconds = 100% - (99%) = 1%
2 seconds = 100% - (99% * 98%) = 3%
3 seconds = 100% - (99% * 98% * 97%) =  6%
4 seconds = 100% - (99% * 98% * 97% * 96%) = 10%
etc.
```

So it snowballs pretty fast. 

With the current rate of change (0.00125 = 8 seconds @ 1%) it looks like this:

singularity_math_code

<details>
<summary>Python Mathplotlib Graph Code</summary>

```python
import matplotlib.pyplot as plt

# Define starting probability and increment (considering 0% chance)
start_probability = -0.01875
increment = 0.00125 #0.00125

# Create time list (up to 2 minutes)
time_list = range(121)  # 120 seconds (2 minutes)

# Calculate cumulative probability for each second
probability_list = []
cumulative_probability = 1
for time in time_list:
  if time == 0:
    cumulative_probability = 1
  else:
    current_probability = max(start_probability + (increment * time), 0)
    cumulative_probability = cumulative_probability * (1 - current_probability)

  probability_list.append(100 - (cumulative_probability * 100))  # Convert to percentage

# Create the plot
plt.figure(figsize=(16, 10))
plt.plot(time_list, probability_list)

# Set labels and title
plt.xlabel("Time (Seconds)")
plt.ylabel("Cumulative Probability (%)")
plt.title("Cumulative Probability of Singularity Event Over Time")

# Adjust y-axis limits to show relevant range
plt.ylim(bottom=0, top=100)  # Set to 0-100 for probability

# Show the plot
plt.grid(True)
plt.show()
```
</detail>

**Comment by zoomachina** [comment]
> Keep in mind the [probability is cumulative](https://blog.minitab.com/en/understanding-statistics/difference-between-probability-and-cumulative-probability). To make it simple if your rate of change was 1% then every second your chance of the event being triggered would be:

if it has a chance to trigger every second then it's ok. I thought the event had a chance to trigger only every 8 seconds, which would mean a 75% chance it won't happen for at least a minute

**Comment by Blacklist897** [comment]
yeah this is kinda shit, not only will sec be forced to keep killing every antag because they get unstoppable but its also "murder so you call the shuttle and then do a longer hijack when everyone on the shuttle is allready dead

**Comment by CCC23** [comment]
> More traitor final objectives that just encourage sec to kill every antag they can before they turn into a nigh unstoppable force. I dislike this PR because it further contributes towards that.

where the hell are you that thats NOT happening I got shot dead in cuffs because I hacked a door and they found a uplink

**Comment by SimplySmiley123** [comment]
> > More traitor final objectives that just encourage sec to kill every antag they can before they turn into a nigh unstoppable force. I dislike this PR because it further contributes towards that.
> 
> where the hell are you that thats NOT happening I got shot dead in cuffs because I hacked a door and they found a uplink

Yeah. You're a guaranteed round ending threat now as traitor.

**Comment by Blacklist897** [comment]
Anyway whats the justification for this anyway? whys the syndicate destroying a random shuttle with a singlo instead of hijacking or anything else

**Comment by nianjiilical** [comment]
i dont honestly see this working without either being disappointing for the traitor or impossible to pull off meaningfully because of the crew

if a traitor rolls any other final objective, they have something cool they can do to cause the round to end, and still have people meaningfully interact with/have to deal with it. if they roll this their two options are to force a shuttle call through boring sabotage (at which point why not just give them a final objective that does it in an actually cool way) or kill everybody (at which point why even place the beacon)

plus shuttles are small, and with how many crewmembers can break into the cockpit with job/tiding tools, i legit dont see how this is going to be allowed to happen if the shuttle isnt depopulated (see above). people will sometimes tolerate antags hijacking because shuttle hijack doesnt actually do anything harmful from a mechanical standpoint, but the minute you put down the beacon everybody who wants to make it to centcom is going to be on your ass

if you managed to fortify the shuttle to the point where you can pull it off with the crew alive then a) you're already at a point where you've basically won your greentext b) you're just destroying the shuttle in a slightly exotic way and not doing anything that you couldnt do just by killing everybody c) you probably just did it by killing everyone in a less fun way than romerol/battlecruiser/resonance cascade

i just dont see how this isnt going to either never go off or just be read by most competent antags as "kill everyone so the shuttle is yours and then kill yourself on it" and that simply isnt a fun final objective to get imo

if this is the level of hoops we have to go through to make singulo shuttle events happen in-round we might as well just keep it admin only

**Comment by Loafin34** [comment]
two questions:
one, can it be defused like a bomb?
two, can you anchor it in place like a bomb (that wrenches can't unwrench?)

**Comment by Exester509** [comment]
I don't like this. Most other final objectives are either instant or rather fast and consistent, the dark matter singularity being the one that takes the longest. Talking about which, we already have a special singularity as a better implemented final objective.

This takes forever to set up, is the easiest to shut down, and the least consequential.

File: code/game/machinery/syndicatebeacon.dm [hunk 1]
```diff
@@ -106,6 +106,98 @@
...
+
+		if(!no_escape_event)
+			var/area/escape_shuttle_area = get_area(src)
+			// beacon must be on the traveling escape shuttle (not a pod)
+			if(istype(escape_shuttle_area, /area/shuttle/escape) && (SSshuttle.emergency.mode == SHUTTLE_ESCAPE) && SSshuttle.emergency.is_in_shuttle_bounds(src))
+				var/obj/docking_port/mobile/port = SSshuttle.emergency
+				no_escape_event = new /datum/shuttle_event/simple_spawner/black_hole/no_escape(port)
+				no_escape_event.beacon = src
+				port.event_list.Add(no_escape_event)
+				no_escape_event.start_up_event(SSshuttle.emergency_escape_time * port.engine_coeff)
```
Comment (Time-Green) [inline code/game/machinery/syndicatebeacon.dm:197]: I added a new add_shuttle_event(typepath) to the port for easier event adding. You'll have to make it return the instance though to add the beacon

File: code/modules/antagonists/traitor/objectives/final_objective/no_escape.dm [hunk 2]
```diff
@@ -0,0 +1,48 @@
+/datum/traitor_objective/ultimate/no_escape
+	name = "Attatch a beacon to the escape shuttle that will attract a singularity to consume everything."
```
Comment (Time-Green) [inline code/modules/antagonists/traitor/objectives/final_objective/no_escape.dm:2]:
```suggestion
	name = "Attach a beacon to the escape shuttle that will attract a singularity to consume everything."
```

**Review by Time-Green — CHANGES_REQUESTED** [review]
Nice work! Sorry for the delay

File: code/__DEFINES/gravity.dm [hunk 3]
```diff
@@ -12,6 +12,34 @@
...
+/// Singularity stage 4 energy requirement
+#define STAGE_FOUR_ENERGY_REQUIREMENT 1000
+/// Singularity stage 5 energy requirement
+#define STAGE_FIVE_ENERGY_REQUIREMENT 2000
+/// Singularity stage 6 energy requirement (also needs to consume a SM shard)
+#define STAGE_SIX_ENERGY_REQUIREMENT 3000
+
+// These values get the median number between two stages to prevent expansion/shrinkage immediately
+/// Singularity stage 1
+#define STAGE_ONE_ENERGY ((STAGE_TWO_ENERGY_REQUIREMENT - STAGE_ONE_ENERGY_REQUIREMENT) / 2 ) + STAGE_ONE_ENERGY_REQUIREMENT
```
[the remark this answers is not in the document]
Comment (timothymtorres) [reply code/__DEFINES/gravity.dm]: Applied with https://github.com/tgstation/tgstation/pull/86796/commits/b6855b52799b68db5fcdfa3a5b315591cfefc100

File: code/game/machinery/syndicatebeacon.dm [hunk 4]
```diff
@@ -106,6 +106,98 @@
...
+
+/obj/machinery/power/singularity_beacon/syndicate/no_escape/screwdriver_act(mob/living/user, obj/item/tool)
+	return
+
+/obj/machinery/power/singularity_beacon/syndicate/no_escape/emp_act(severity)
+	. = ..()
+	if(machine_stat & (NOPOWER|BROKEN) || . & EMP_PROTECT_CONTENTS)
+		return
+	if(cell)
+		cell.emp_act(severity)
```
[the remark this answers is not in the document]
Comment (timothymtorres) [reply code/game/machinery/syndicatebeacon.dm]: Applied with https://github.com/tgstation/tgstation/pull/86796/commits/8a4aa0e171c47190e0ee892b962fab134e43e062

File: code/modules/antagonists/traitor/objectives/final_objective/no_escape.dm (hunk 2 above)
Comment (timothymtorres) [reply code/modules/antagonists/traitor/objectives/final_objective/no_escape.dm]: Applied with https://github.com/tgstation/tgstation/pull/86796/commits/25f8fabf666065f8aebe287749149f5d4447dae8

File: code/game/machinery/syndicatebeacon.dm (hunk 1 above)
Comment (timothymtorres) [reply code/game/machinery/syndicatebeacon.dm]: Done.

**Comment by Time-Green** [comment]
This is going to be so peak

**Review by Time-Green — APPROVED** [review]

**Comment by Time-Green** [comment]
@MrMelbert I forgot to remove our names from the PR bodies so uhh enjoy being pinged by every downstream

**Pull request merged**